File: src/common/web/data/entities/project/features/ProjectMetadataFeature.ts
import { Type } from "class-transformer";

import { LayoutPropertyObject } from "../../../../ui/components/propertyeditor/PropertyObjectStore";
import { type ProfileID } from "../../../../ui/components/propertyeditor/PropertyProfile";
import { ProjectFeature, type ProjectFeatureID } from "./ProjectFeature";

/**
 * The project metadata type.
 *
 *
 */
export type ProjectMetadata = LayoutPropertyObject[];

/**
 * Data class for the metadata project feature.
 */
export class ProjectMetadataFeature extends ProjectFeature {
    public static readonly FeatureID: ProjectFeatureID = "project_metadata";

    // @ts-ignore
    @Type(() => LayoutPropertyObject)
    public readonly metadata: ProjectMetadata;

    public constructor(metadata: ProjectMetadata = [], enabledProfiles: ProfileID[] = []) {
        super(enabledProfiles);

        this.metadata = metadata;
    }

    public get featureID(): ProjectFeatureID {
        return ProjectMetadataFeature.FeatureID;
    }
}

File: src/common/py/data/entities/project/features/project_metadata_feature.py
import typing
from dataclasses import dataclass, field

from dataclasses_json import dataclass_json

from ...properties.properties_object import PropertyObject
from .project_feature import ProjectFeature, ProjectFeatureID

ProjectMetadata = typing.List[PropertyObject]


@dataclass_json
@dataclass
class ProjectMetadataFeature(ProjectFeature):
    """
    Data class for the project metadata project feature.
    """

    feature_id: typing.ClassVar[ProjectFeatureID] = "project_metadata"

    metadata: ProjectMetadata = field(default_factory=list)
